Allow replacing visualizations on dashboard

friol opened the pull request:
## Summary

As asked in Issue #43900, this PR adds the "Replace visualization" function to the Edit Menu when a Dashboard is in edit mode. The new visualization is chosen from a flyout, and replaces the old one with the same position and dimensions.

![replaceViewScreenshot](https://user-images.githubusercontent.com/4502137/64473514-23446680-d168-11e9-9e99-bcfad2ee36c8.png)

elasticmachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

elasticmachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

elasticmachine commented:
Pinging @elastic/kibana-app

TinaHeiligers commented:
jenkins test this

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3204/)

friol commented:
jenkins test this

friol commented:
retest

friol commented:
@TinaHeiligers : can you please run the build again?

timroes commented:
@stacey-gammon Since the embeddable layer is still rather new, could we perhaps bother you for a review here?

timroes commented:
Jenkins, test this

stacey-gammon commented:
This is awesome!  You might have to update some things after https://github.com/elastic/kibana/pull/44707 is merged, hopefully today.  I am out tomorrow and Friday and I'm not sure I'll be able to finish the review today but if not I'll get back around to it early next week.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3661/)

friol commented:
@stacey-gammon : the type_check script fails for this error:

ERROR x-pack failed
      src/legacy/core_plugins/dashboard_embeddable_container/public/np_ready/public/index.ts:20:42 - error TS2307: Cannot find module 'kibana/public'.
      20 import { PluginInitializerContext } from 'kibana/public';

But the error is not in files I've touched. The same line of code is also on master:

[https://github.com/elastic/kibana/blob/master/src/legacy/core_plugins/dashboard_embeddable_container/public/np_ready/public/index.ts](https://github.com/elastic/kibana/blob/master/src/legacy/core_plugins/dashboard_embeddable_container/public/np_ready/public/index.ts)

timroes commented:
Jenkins, test this

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/4075/)

friol commented:
@timroes : error is not in my code, is it?

stacey-gammon commented:
@friol , it's probably that something in jest is now importing that file that wasn't before. Those imports don't work in jenkins unfortunately so the recommendation is still relative file paths (discussed here: https://github.com/elastic/kibana/issues/40446). We eventually we want to move to support these short file imports, and they are still scattered about the code because they don't always cause failures.  Short term fix is to just switch to a relative file path import.

friol commented:
> @friol , it's probably that something in jest is now importing that file that wasn't before. Those imports don't work in jenkins unfortunately so the recommendation is still relative file paths (discussed here: #40446). We eventually we want to move to support these short file imports, and they are still scattered about the code because they don't always cause failures. Short term fix is to just switch to a relative file path import.

@stacey-gammon : ok, so, exactly, what should I do?

stacey-gammon commented:
Change the line in      src/legacy/core_plugins/dashboard_embeddable_container/public/np_ready/public/index.ts from

`import { PluginInitializerContext } from 'kibana/public';` to

`import { PluginInitializerContext } from '../../../../../../core/public';`

friol commented:
> Change the line in src/legacy/core_plugins/dashboard_embeddable_container/public/np_ready/public/index.ts from
> 
> `import { PluginInitializerContext } from 'kibana/public';` to
> 
> `import { PluginInitializerContext } from '../../../../../../core/public';`

@stacey-gammon : Ah, ok, I didn't understand I had to modify files outside my code.
Will do it this evening.

friol commented:
@stacey-gammon : done.

stacey-gammon commented on src/legacy/core_plugins/embeddable_api/public/np_ready/public/lib/panel/panel_header/panel_actions/add_panel/change_view_flyout.tsx:
```diff
@@ -0,0 +1,137 @@
+    const originalPanels = this.props.container.getInput().panels;
+    const filteredPanels = { ...originalPanels };
+
+    const nnw = (filteredPanels[this.props.viewToRemove.id] as DashboardPanelState).gridData.w;
```
Ah, I just realized this action is coupled to the dashboard container.  In that case it should actually not be a part of this plugin but the `dashboard_container_embeddable` one.  We don't want it to show up for embeddables that aren't in a dashboard, and this plugin doesn't depend on that one (other way around actually).

When you do that you'll have to attach this action to the `CONTEXT_MENU_TRIGGER`, similar to the other action dashboard registers:

```
    uiActions.attachAction(CONTEXT_MENU_TRIGGER, expandPanelAction.id);
```

friol replied:
Sorry, it's not clear to me.
All the other actions (add panel, inspect panel, etc.) are under the embeddable_api subtree.
I just placed my action with those other ones (maybe wrongly).

stacey-gammon replied:
The actions in the embeddable_api are relevant for all embeddables, in any container.  If an action is only applicable to a certain type, it should be in that plugin.  If you check out `ExpandPanelAction` inside the `dashboard_embeddable_container` plugin, you can use that as an example.  The way it is now there is a circular dependency - you are importing code from the dashboard embeddable plugin, but that plugin needs code from this plugin.  This won't work when we migrate all these plugins to the "new platform".

lmk if you need some more help on how to migrate this action over!

friol replied:
Hope to have done this right...

stacey-gammon commented on src/legacy/core_plugins/embeddable_api/public/np_ready/public/lib/panel/panel_header/panel_actions/change_view_action.ts:
```diff
@@ -0,0 +1,101 @@
+      hasExpandedPanelInput(embeddable.parent) &&
+      embeddable.parent.getInput().expandedPanelId === embeddable.id;
+
+    return Boolean(
```
Once in dashboard plugin, you'll want to do a check in here to make sure the container type is a dashboard.  It'd be great if we could do it generally but the logic is specific to dashboard, and I don't think there is a good way to do it generally.

friol replied:
Ok, done.

stacey-gammon commented on src/legacy/core_plugins/embeddable_api/public/np_ready/public/lib/panel/panel_header/panel_actions/change_view_action.ts:
```diff
@@ -0,0 +1,101 @@
+  public readonly id = CHANGE_VIEW_ACTION;
+  public order = 11;
+
+  constructor(
```
nit: I think the pattern of taking in an object is easier to follow than a list of parameters, then order doesn't matter.

friol replied:
Order influences the position of the new item ("Replace visualization") in the contextual menu.
I don't understand fully what you mean here.

stacey-gammon replied:
Sorry I meant order of the parameters, not the parameter named `order`. :)   When passing in a lot of arguments, a reader can know what they are, and a caller doesn't need to worry about which parameter is first, second third.  E.g. they can do :

```
new ChangeViewAction({ getFactory, getAllFactories });
```
or
```
new ChangeViewAction({ getAllFactories, getFactory});
```

So in situations like this, I find a single object being passed into the constructor easier to read.  It's just a nit though, feel free to leave as is if you like it this way better.

friol replied:
I understand what you're saying, but I'm just copying the code that is present somewhere else, for example in add_panel_action.ts.

stacey-gammon replied:
ah, which I wrote, so I guess I don't follow my own advice, whoops! :D  Fine to leave as is.

friol commented on src/legacy/core_plugins/embeddable_api/public/np_ready/public/lib/panel/panel_header/panel_actions/add_panel/change_view_flyout.tsx:
```diff
@@ -0,0 +1,137 @@
+  EuiFlyoutBody,
+  EuiFlyoutHeader,
+  // @ts-ignore
+  EuiSuperSelect,
```
Ok, done.

friol commented on src/legacy/core_plugins/embeddable_api/public/np_ready/public/lib/panel/panel_header/panel_actions/change_view_action.ts:
```diff
@@ -0,0 +1,101 @@
+  }
+
+  public async execute({ embeddable }: ActionContext) {
+    if (embeddable.parent) {
```
Ok, done.

stacey-gammon commented:
@elasticmachine merge upstream

stacey-gammon commented:
retest

stacey-gammon commented on src/legacy/core_plugins/dashboard_embeddable_container/public/np_ready/public/lib/actions/change_view_flyout.tsx:
```diff
@@ -0,0 +1,135 @@
+  IEmbeddable,
+  EmbeddableInput,
+  EmbeddableOutput,
+} from '../../../../../../embeddable_api/public/np_ready/public/lib/embeddables';
```
```suggestion
} from '../../../../../../embeddable_api/public/np_ready/public';
```

plugins are only allowed to import top level exports from other plugins, so I think you will need to clean up a couple of the import paths to get past the type errors

friol replied:
Done.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/1511/)

stacey-gammon commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/2486/)
* Commit: 7d8cd897afeb021faa29b9854c71a11117eb5433

stacey-gammon commented:
@friol, just a few more changes to fix the i18n errors:

```
ERROR  I18N ERROR  Error in src/legacy/core_plugins/dashboard_embeddable_container/public/np_ready/public/lib/actions/change_view_action.tsx
      Error: Expected "embeddableApi.panel.removePanel.replaceView" id to have "dashboardEmbeddableContainer" namespace. See .i18nrc.json for the list of supported namespaces.
ERROR  I18N ERROR  Error in src/legacy/core_plugins/dashboard_embeddable_container/public/np_ready/public/lib/actions/change_view_flyout.tsx
      Error: Expected "embeddableApi.addPanel.savedObjectAddedToContainerSuccessMessageTitle" id to have "dashboardEmbeddableContainer" namespace. See .i18nrc.json for the list of supported namespaces.
ERROR  I18N ERROR  Error in src/legacy/core_plugins/dashboard_embeddable_container/public/np_ready/public/lib/actions/change_view_flyout.tsx
      Error: Expected "embeddableApi.addPanel.noMatchingObjectsMessage" id to have "dashboardEmbeddableContainer" namespace. See .i18nrc.json for the list of supported namespaces.
```

Just change the id from `embeddableApi` to `dashboardEmbeddableContainer`.

Looks like all other tests passed! 🤞 

Once those id changes are passed, I'll run the tests again and once green, give it a final pass.

friol commented:
@stacey-gammon : done!

stacey-gammon commented:
retest

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/2600/)
* Commit: 736e3e4219477f8aed66c3138fdaa42fbc605275

stacey-gammon commented:
woohoo!  My final review will have to wait till Monday, I'm heading out for the day, but feel free to send me a ping if I don't review by EOD Monday, just in case I forget!  But I'll try not to!

stacey-gammon commented:
@elasticmachine merge upstream

stacey-gammon commented:
retest

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/2803/)
* Commit: 736e3e4219477f8aed66c3138fdaa42fbc605275

stacey-gammon commented on src/legacy/core_plugins/dashboard_embeddable_container/public/np_ready/public/lib/actions/change_view_action.tsx:
```diff
@@ -0,0 +1,88 @@
+  }
+
+  public async isCompatible({ embeddable }: ActionContext) {
+    return Boolean(embeddable.parent && isDashboard(embeddable.parent));
```
should check for edit mode as well, we don't want it showing up in view mode.

friol replied:
Ok, right.

stacey-gammon commented on src/legacy/core_plugins/dashboard_embeddable_container/public/np_ready/public/lib/actions/change_view_flyout.tsx:
```diff
@@ -0,0 +1,138 @@
+
+interface Props {
+  container: IContainer;
+  sof: React.ComponentType<any>;
```
Can you spell this out as `savedObjectFinder`?

friol replied:
Ok, done.

stacey-gammon commented on src/legacy/core_plugins/dashboard_embeddable_container/public/np_ready/public/lib/actions/change_view_flyout.tsx:
```diff
@@ -0,0 +1,138 @@
+}
+
+export class ChangeViewFlyout extends React.Component<Props> {
+  private lastToast: any;
```
```suggestion
 private lastToast?: Toast;
```

friol replied:
Ok done.

stacey-gammon commented on src/legacy/core_plugins/dashboard_embeddable_container/public/np_ready/public/lib/actions/change_view_flyout.tsx:
```diff
@@ -0,0 +1,138 @@
+import { i18n } from '@kbn/i18n';
+import React from 'react';
+
+import { NotificationsStart } from 'src/core/public';
```
```suggestion
import { NotificationsStart, Toast } from 'src/core/public';
```

friol replied:
Done.

stacey-gammon commented on src/legacy/core_plugins/dashboard_embeddable_container/public/np_ready/public/lib/actions/change_view_flyout.tsx:
```diff
@@ -0,0 +1,138 @@
+      />
+    );
+
+    const vtr = 'Replace view ' + this.props.viewToRemove.getTitle() + ' with:';
```
nit
```suggestion
    const viewToReplace = 'Replace view ' + this.props.viewToRemove.getTitle() + ' with:';
```

friol replied:
Ok, done.

stacey-gammon commented on src/legacy/core_plugins/dashboard_embeddable_container/public/np_ready/public/lib/actions/open_change_view_flyout.tsx:
```diff
@@ -0,0 +1,56 @@
+export async function openChangeViewFlyout(options: {
+  embeddable: IContainer;
+  core: CoreStart;
+  savedobjectfinder: React.ComponentType<any>;
```
```suggestion
  savedObjectFinder: React.ComponentType<any>;
```

friol replied:
Done.

friol commented:
> Looking good! Just a few minor things, and then tests should be the last thing needed. One functional, the rest jest tests for decent code coverage would be good. lmk if you need any guidance for writing tests. Thanks again for your continued effort on this feature!

Ok, will try to do this tomorrow.

friol commented:
@stacey-gammon : ok, "done". 
For the tests, I just copied expand_panel_action.test.tsx and I'm missing a proper test of the functionality (the commented part). Need some help on this. Thanks.

stacey-gammon commented:
@elasticmachine merge upstream

stacey-gammon commented on src/legacy/core_plugins/dashboard_embeddable_container/public/np_ready/public/lib/actions/change_view_action.test.tsx:
```diff
@@ -0,0 +1,135 @@
+});
+
+/*
+test('Sets the embeddable expanded panel id on the parent', async () => {
```
Should this be uncommented?  Or deleted?

friol replied:
See my comment above...

stacey-gammon commented:
Thanks @friol !  Someone from the @elastic/kibana-app team should be picking this up soon to help with tests.  It's got an LGTM from me for the embeddable part of it.

Oh one thing I did think of though this weekend - it should probably be displayed as "Replace panel" not "Replace visualization" since you could swap it out with different types of embeddables, and the term "visualization" is specific to the one type.  We also have "map" and "saved search", may eventually have "input control", and others.

timroes commented:
@elasticsearch merge upstream

friol commented:
> Thanks @friol ! Someone from the @elastic/kibana-app team should be picking this up soon to help with tests. It's got an LGTM from me for the embeddable part of it.
> 
> Oh one thing I did think of though this weekend - it should probably be displayed as "Replace panel" not "Replace visualization" since you could swap it out with different types of embeddables, and the term "visualization" is specific to the one type. We also have "map" and "saved search", may eventually have "input control", and others.

@stacey-gammon : I renamed the menu item to "Replace panel". To be honest, one would have to change all the file names and all the object names to "change_panel_something"... but I hope you won't need that :)

nickofthyme commented:
@friol Thanks for jumping on this so quickly I thought of this feature just a few months ago!

I updated the naming to `replacePanel` just to be consistent across the board.

I couldn't find any issue with the test you pointed out, tbd is the ci passes 🤞.

I also added some functional tests to your PR. Please let me know if you have any comments witht the changes and feel free to make any changes you'd like.

nickofthyme commented:
retest please

nickofthyme commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/5628/)
* Commit: 8034d10bbeca1273eca26766db2290a6475624ad

nickofthyme commented:
retest

friol commented:
> @friol Thanks for jumping on this so quickly I thought of this feature just a few months ago!
> 
> I updated the naming to `replacePanel` just to be consistent across the board.
> 
> I couldn't find any issue with the test you pointed out, tbd if the ci passes 🤞.
> 
> I also added some functional tests to your PR. Please let me know if you have any comments witht the changes and feel free to make any changes you'd like.

@nickofthyme : simply great! You had the patience to rename all the objects to "replacePanel*" :)
Many thanks - I don't have any comments, it's ok to me.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/5662/)
* Commit: e8dceeaddd47ed48db4a2b3b962b9fff92c744c3

nickofthyme commented:
@elasticmachine merge upstream

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/6445/)
* Commit: 61b2ade0b46b120cb78efa8dc93cf80a41d5e555

friol commented:
@nickofthyme : tell me we can finally merge this one! It's 1 month and 15 days old :)

nickofthyme commented:
@friol Done! Thanks again for all the help!

friol commented:
@nickofthyme : there is another PR from me, dated 28 august, that got "lost" -> #35379 .
Would you mind reviewing it? I think that functionality would be quite useful.
Thanks

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/6715/)
* Commit: 61b2ade0b46b120cb78efa8dc93cf80a41d5e555

nickofthyme commented:
@friol I'll take a look at #35379.

FYI this feature will be in the `7.6.0` release, just missed `7.5.0` 😞

friol commented:
> @friol I'll take a look at #35379.
> 
> FYI this feature will be in the `7.6.0` release, just missed `7.5.0` 😞

@nickofthyme : thanks!
